packages/mobile/src/screens/tip-artist-screen/SendTipScreen.tsx

```diff
@@ -50,12 +64,112 @@ const parseToBNWei = (tipAmount: StringAudio) => {
 export const SendTipScreen = () => {
   const styles = useStyles()
   const [tipAmount, setTipAmount] = useState('')
-  const accountBalance = useSelectorWeb(getAccountBalance)
+  const accountBalance = (useSelectorWeb(getAccountBalance) ??
+    new BN('0')) as BNWei
   const navigation = useNavigation<TipArtistNavigationParamList>()
   const dispatchWeb = useDispatchWeb()
 
-  const hasInsufficientBalance =
-    accountBalance && parseToBNWei(tipAmount).gt(accountBalance)
+  const account = useSelectorWeb(getAccountUser)
+  const supportersMap = useSelectorWeb(getOptimisticSupporters)
+  const supportingMap = useSelectorWeb(getOptimisticSupporting)
+  const receiver = useSelectorWeb(getSendUser)
+  const [
+    amountToTipToBecomeTopSupporter,
+    setAmountToTipToBecomeTopSupporter
+  ] = useState<Nullable<BNWei>>(null)
+  const [supportingAmount, setSupportingAmount] = useState<Nullable<StringWei>>(
+    null
+  )
+  const [topSupporter, setTopSupporter] = useState<Nullable<Supporter>>(null)
+  const [isFirstSupporter, setIsFirstSupporter] = useState(false)
+
+  const tipAmountWei = parseToBNWei(tipAmount)
+  const hasInsufficientBalance = tipAmountWei.gt(accountBalance)
+
+  /**
+   * Get supporting info if current user is already supporting receiver
+   * so that the already supported amount can be used to determine
+   * how much is left to tip to become top supporter
+   */
+  useEffect(() => {
+    if (!account || !receiver) return
+    if (supportingAmount) return
+
+    const supportingForAccount = supportingMap[account.user_id] ?? null
+    const accountSupportingReceiver =
+      supportingForAccount?.[receiver.user_id] ?? null
+    if (accountSupportingReceiver) {
+      setSupportingAmount(accountSupportingReceiver.amount)
+    } else {
+      dispatchWeb({
+        type: MessageType.FETCH_USER_SUPPORTER,
+        currentUserId: account.user_id,
+        userId: receiver.user_id,
+        supporterUserId: account.user_id
+      })
+    }
+  }, [account, receiver, supportingAmount, supportingMap, dispatchWeb])
+
+  /**
+   * Get user who is top supporter to later check whether it is
+   * not the same as the current user
+   */
+  useEffect(() => {
+    if (!receiver) return
+
+    const supportersForReceiver = supportersMap[receiver.user_id] ?? {}
+    const rankedSupportersList = Object.keys(supportersForReceiver)
```

Comment:
Might be worth making a utility function to convert the map to a ranked list, as I think we do this in a few places now?

Reply:
true that